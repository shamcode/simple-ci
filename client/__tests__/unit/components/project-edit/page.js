import { DI } from 'sham-ui';
import Page from '../../../../src/components/project-edit/page.sfc';
import * as directives from 'sham-ui-directives';
import hrefto from 'sham-ui-router/href-to';
import renderer from 'sham-ui-test-helpers';

const flushPromises = () => new Promise( resolve => setImmediate( resolve ) );

afterEach( () => {
    DI.bind( 'store', null );
    DI.bind( 'router', null );
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'renders correctly', async() => {
    expect.assertions( 5 );

    const storeMock = jest.fn().mockReturnValue( Promise.resolve( {
        project: {
            id: 1,
            name: 'Test',
            cwd: '/tmp/'
        }
    } ) );
    DI.bind( 'store', {
        getProjectById: storeMock
    } );
    DI.bind( 'router', {
        lastRouteResolved: jest.fn().mockReturnValueOnce( { params: { id: 1 } } ),
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( Page, {
        directives: {
            ...directives,
            hrefto
        }
    } );
    const { component: { container } } = meta;
    expect( storeMock.mock.calls[ 0 ][ 0 ] ).toBe( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();

    await flushPromises();
    expect( container.querySelector( '[name="name"]' ).value ).toBe( 'Test' );
    expect( container.querySelector( '[name="cwd"]' ).value ).toBe( '/tmp/' );
    expect( meta.toJSON() ).toMatchSnapshot();
    await flushPromises();
} );

it( 'update fail', async() => {
    expect.assertions( 2 );

    const updateProjectMock = jest.fn();
    DI.bind( 'store', {
        getProjectById: jest.fn().mockReturnValue( Promise.resolve( {
            project: {
                id: 1,
                name: 'Test',
                cwd: '/tmp/'
            }
        } ) ),
        updateProject: updateProjectMock
    } );
    DI.bind( 'router', {
        lastRouteResolved: jest.fn().mockReturnValueOnce( { params: { id: 1 } } ),
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( Page, {
        directives: {
            ...directives,
            hrefto
        }
    } );
    const { component: { container } } = meta;
    await flushPromises();


    updateProjectMock.mockReturnValueOnce( Promise.reject() );
    container.querySelector( '[type="submit"]' ).click();
    expect( updateProjectMock.mock.calls ).toHaveLength( 1 );

    await flushPromises();
    expect( meta.toJSON() ).toMatchSnapshot();
} );
