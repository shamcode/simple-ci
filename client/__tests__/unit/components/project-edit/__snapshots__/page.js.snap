// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "Constructor": "ProjectEditPage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "dataLoaded": false,
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [],
    "project": Object {},
    "types": Array [],
    "updateProject": [Function],
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\"> Projects </a>
  </li>
  <!--if-->
  <li></li>
</ul>
<!--BreadCrumb-->
<!--breadcrumbs-->
<!--if-->
<!--WithoutFlushLoader-->
<!--if-->
<!--PageContent-->
",
}
`;

exports[`renders correctly 2`] = `
Object {
  "Constructor": "ProjectEditPage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "dataLoaded": true,
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [],
    "project": Object {
      "cwd": "/tmp/",
      "id": 1,
      "name": "Test",
    },
    "types": Array [],
    "updateProject": [Function],
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\"> Projects </a>
  </li>
  <li>
    <a href=\\"undefined\\">Test</a>
  </li>
  <!--if-->
  <!--if-->
  <li>Edit</li>
</ul>
<!--BreadCrumb-->
<!--breadcrumbs-->
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Update</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
<!--content-->
<!--if-->
<!--PageContent-->
",
}
`;

exports[`update fail 1`] = `
Object {
  "Constructor": "ProjectEditPage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "dataLoaded": true,
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [
      "Update project fail!",
    ],
    "project": Object {
      "cwd": "/tmp/",
      "id": 1,
      "name": "Test",
    },
    "types": Array [],
    "updateProject": [Function],
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\"> Projects </a>
  </li>
  <li>
    <a href=\\"undefined\\">Test</a>
  </li>
  <!--if-->
  <!--if-->
  <li>Edit</li>
</ul>
<!--BreadCrumb-->
<!--breadcrumbs-->
<ul class=\\"errors-panel\\">
  <li>Update project fail!</li>
</ul>
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Update</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
<!--content-->
<!--if-->
<!--PageContent-->
",
}
`;
